File: scripts/generate-og.mjs
// Generate OG image using Node.js built-in APIs
// Creates a 1200x630 PNG with dark background, logo, and text

import { readFileSync, writeFileSync } from "fs";
import { join, dirname } from "path";
import { fileURLToPath } from "url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const projectRoot = join(__dirname, "..");
const merkRoot = join(projectRoot, "..");

// We'll create an SVG and convert it to a simple approach
// Since we don't have sharp/canvas installed, create an SVG-based OG image
// Actually, let's use a different approach: create the image using Next.js OG image generation

// For now, let's create a simple HTML file that can be screenshot'd,
// or better yet, let's check if sharp is available
let sharp;
try {
  sharp = (await import("sharp")).default;
} catch {
  console.log("sharp not available, trying alternative...");
  // Create a minimal PNG using raw buffer manipulation
  // This is a fallback approach
  sharp = null;
}

if (sharp) {
  const logoPath = join(merkRoot, "logo", "merkaba_digital.png");
  const outputPath = join(projectRoot, "public", "og-image.png");

  const logo = await sharp(logoPath)
    .resize({ height: 300, withoutEnlargement: true })
    .toBuffer();

  const logoMeta = await sharp(logo).metadata();
  const logoW = logoMeta.width;
  const logoH = logoMeta.height;
  const logoX = Math.round((1200 - logoW) / 2);
  const logoY = Math.round((630 - logoH) / 2) - 60;

  const textSvg = `<svg width="1200" height="630">
    <text x="600" y="${logoY + logoH + 70}"
      text-anchor="middle"
      font-family="Helvetica, Arial, sans-serif"
      font-size="36"
      fill="white">Generative Engine Optimization</text>
  </svg>`;

  const result = await sharp({
    create: {
      width: 1200,
      height: 630,
      channels: 4,
      background: { r: 11, g: 14, b: 26, alpha: 1 },
    },
  })
    .composite([
      { input: logo, top: logoY, left: logoX },
      { input: Buffer.from(textSvg), top: 0, left: 0 },
    ])
    .png()
    .toFile(outputPath);

  console.log("OG image saved to", outputPath);
} else {
  // Fallback: create SVG-based image and save as SVG, or use a minimal PNG approach
  // Since neither sharp nor canvas is available, create the image via an alternative

  // Try using the Python approach
  const { execSync } = await import("child_process");
  try {
    execSync("python3 " + join(__dirname, "generate-og.py"), { stdio: "inherit" });
  } catch {
    console.error("Could not generate OG image. Please run manually:");
    console.error("  python3 scripts/generate-og.py");
    console.error("  OR install sharp: npm install sharp && node scripts/generate-og.mjs");
  }
}

File: scripts/generate-og.py
#!/usr/bin/env python3
"""Generate OG image for Merkaba Digital website."""

from PIL import Image, ImageDraw, ImageFont
import os

WIDTH = 1200
HEIGHT = 630
BG_COLOR = (11, 14, 26)  # #0B0E1A
TEXT_COLOR = (255, 255, 255)

script_dir = os.path.dirname(os.path.abspath(__file__))
project_root = os.path.dirname(script_dir)
merkaba_root = os.path.dirname(project_root)

logo_path = os.path.join(merkaba_root, "logo", "merkaba_digital.png")
output_path = os.path.join(project_root, "public", "og-image.png")

img = Image.new("RGB", (WIDTH, HEIGHT), BG_COLOR)
draw = ImageDraw.Draw(img)

# Load and center the logo
logo = Image.open(logo_path).convert("RGBA")

# Scale logo to fit nicely (max 300px height, keeping aspect ratio)
max_logo_h = 300
ratio = max_logo_h / logo.height
new_w = int(logo.width * ratio)
new_h = int(logo.height * ratio)
logo = logo.resize((new_w, new_h), Image.LANCZOS)

# Paste logo centered, offset up a bit to leave room for text
logo_x = (WIDTH - new_w) // 2
logo_y = (HEIGHT - new_h) // 2 - 60

# Create a temporary RGBA image for compositing
tmp = Image.new("RGBA", (WIDTH, HEIGHT), BG_COLOR + (255,))
tmp.paste(logo, (logo_x, logo_y), logo)
img = tmp.convert("RGB")
draw = ImageDraw.Draw(img)

# Text below logo
text = "Generative Engine Optimization"

# Try to use a nice font, fall back to default
font_size = 36
try:
    # Try common macOS fonts
    for font_name in [
        "/System/Library/Fonts/Helvetica.ttc",
        "/System/Library/Fonts/SFNSMono.ttf",
        "/Library/Fonts/Arial.ttf",
    ]:
        if os.path.exists(font_name):
            font = ImageFont.truetype(font_name, font_size)
            break
    else:
        font = ImageFont.load_default()
except Exception:
    font = ImageFont.load_default()

bbox = draw.textbbox((0, 0), text, font=font)
text_w = bbox[2] - bbox[0]
text_x = (WIDTH - text_w) // 2
text_y = logo_y + new_h + 40

draw.text((text_x, text_y), text, fill=TEXT_COLOR, font=font)

img.save(output_path, "PNG")
print(f"OG image saved to {output_path}")
